File: stack_queue/queue.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <stdexcept> // For exceptions

/**********************************************
 * CATALOG OF QUEUE FUNCTIONS
 **********************************************
 * 
 * 1. `Queue()`:
 *    Initializes an empty queue.
 * 
 * 2. `bool isEmpty() const`:
 *    Checks if the queue is empty. Returns true or false.
 * 
 * 3. `int size() const`:
 *    Returns the number of elements in the queue.
 * 
 * 4. `void enqueue(int value)`:
 *    Adds an element to the rear of the queue.
 * 
 * 5. `int dequeue()`:
 *    Removes and returns the front element of the queue.
 *    Throws an error if the queue is empty.
 * 
 * 6. `int peek() const`:
 *    Returns the front element without removing it.
 *    Throws an error if the queue is empty.
 * 
 * 7. `Queue merge(const Queue& other) const`:
 *    Merges two queues into a new queue.
 * 
 * 8. `void reverse()`:
 *    Reverses the elements of the queue in place.
 * 
 * 9. `Queue copy() const`:
 *    Creates and returns a copy of the queue.
 * 
 * 10. `void sort()`:
 *     Sorts the elements of the queue in ascending order.
 * 
 * 11. `Queue slice(int start, int end) const`:
 *     Creates and returns a new queue containing elements
 *     from index `start` to `end - 1`.
 **********************************************/

class Queue {
private:
    std::vector<int> data; // Vector for dynamic storage
    int frontIdx;          // Index of the front element
    int size;              // Current number of elements

public:
    // Constructor
    Queue() : frontIdx(0), size(0) {}

    // Check if the queue is empty
    bool isEmpty() const {
        return size == 0;
    }

    // Get the size of the queue
    int getSize() const {
        return size;
    }

    // Add an element to the rear of the queue
    void enqueue(int value) {
        if (frontIdx + size >= (int)data.size()) {
            data.push_back(value);
        } else {
            data[frontIdx + size] = value;
        }
        size++;
    }

    // Remove and return the front element of the queue
    int dequeue() {
        if (isEmpty()) {
            throw std::runtime_error("Queue underflow error!");
        }
        int value = data[frontIdx];
        frontIdx++;
        size--;
        if (size == 0) {
            frontIdx = 0; // Reset indices for empty queue
        }
        return value;
    }

    // Get the front element without removing it
    int peek() const {
        if (isEmpty()) {
            throw std::runtime_error("Queue is empty!");
        }
        return data[frontIdx];
    }

    // Merge with another queue
    Queue merge(const Queue& other) const {
        Queue mergedQueue;
        for (int i = 0; i < size; i++) {
            mergedQueue.enqueue(data[frontIdx + i]);
        }
        for (int i = 0; i < other.size; i++) {
            mergedQueue.enqueue(other.data[other.frontIdx + i]);
        }
        return mergedQueue;
    }

    // Reverse the queue
    void reverse() {
        std::reverse(data.begin() + frontIdx, data.begin() + frontIdx + size);
    }

    // Copy the queue
    Queue copy() const {
        Queue copiedQueue;
        for (int i = 0; i < size; i++) {
            copiedQueue.enqueue(data[frontIdx + i]);
        }
        return copiedQueue;
    }

    // Sort the queue
    void sort() {
        std::sort(data.begin() + frontIdx, data.begin() + frontIdx + size);
    }

    // Slice the queue
    Queue slice(int start, int end) const {
        if (start < 0 || end > size || start >= end) {
            throw std::runtime_error("Invalid slice indices!");
        }
        Queue slicedQueue;
        for (int i = start; i < end; i++) {
            slicedQueue.enqueue(data[frontIdx + i]);
        }
        return slicedQueue;
    }

    // Display the elements of the queue (for testing)
    void display() const {
        for (int i = 0; i < size; i++) {
            std::cout << data[frontIdx + i] << " ";
        }
        std::cout << std::endl;
    }
};

// Main function for testing
int main() {
    Queue q1, q2;

    // Enqueue elements into q1
    q1.enqueue(10);
    q1.enqueue(20);
    q1.enqueue(30);

    // Enqueue elements into q2
    q2.enqueue(40);
    q2.enqueue(50);
    q2.enqueue(60);

    // Merge queues
    Queue mergedQueue = q1.merge(q2);
    std::cout << "Merged queue: ";
    mergedQueue.display();

    // Reverse q1
    q1.reverse();
    std::cout << "Reversed q1: ";
    q1.display();

    // Copy q2
    Queue copiedQueue = q2.copy();
    std::cout << "Copied q2: ";
    copiedQueue.display();

    // Sort q2
    q2.enqueue(25);
    q2.sort();
    std::cout << "Sorted q2: ";
    q2.display();

    // Slice q2
    Queue slicedQueue = q2.slice(1, 4);
    std::cout << "Sliced q2 (1 to 4): ";
    slicedQueue.display();

    return 0;
}

File: stack_queue/queue.py
class Queue:
    """
    CATALOG OF QUEUE FUNCTIONS:
    1. __init__():
    Initializes an empty queue.

    2. is_empty() -> bool:
    Checks if the queue is empty. Returns True or False.

    3. size() -> int:
    Returns the number of elements in the queue.

    4. enqueue(value: int):
    Adds an element to the rear of the queue.

    5. dequeue() -> int:
    Removes and returns the front element of the queue.
    Raises an error if the queue is empty.

    6. peek() -> int:
    Returns the front element without removing it.
    Raises an error if the queue is empty.

    7. merge(other: "Queue") -> "Queue":
    Merges two queues into a new queue.

    8. reverse():
    Reverses the elements of the queue in place.

    9. copy() -> "Queue":
    Creates and returns a copy of the queue.

    10. sort():
        Sorts the elements of the queue in ascending order.

    11. slice(start: int, end: int) -> "Queue":
        Creates and returns a new queue containing elements
        from index `start` to `end - 1`.
    """
    
    def __init__(self):
        """Initialize an empty queue."""
        self.data = []
        self.size = 0

    def is_empty(self):
        """Check if the queue is empty."""
        return self.size == 0

    def size(self):
        """Return the number of elements in the queue."""
        return self.size

    def enqueue(self, value):
        """Add an element to the rear of the queue."""
        self.data.append(value)
        self.size += 1

    def dequeue(self):
        """Remove and return the front element of the queue."""
        if self.is_empty():
            raise RuntimeError("Queue underflow error!")
        value = self.data.pop(0)
        self.size -= 1
        return value

    def peek(self):
        """Return the front element without removing it."""
        if self.is_empty():
            raise RuntimeError("Queue is empty!")
        return self.data[0]

    def merge(self, other):
        """Merge with another queue and return a new queue."""
        merged_queue = Queue()
        merged_queue.data = self.data + other.data
        merged_queue.size = self.size + other.size
        return merged_queue

    def reverse(self):
        """Reverse the elements of the queue in place."""
        self.data.reverse()

    def copy(self):
        """Create and return a copy of the queue."""
        copied_queue = Queue()
        copied_queue.data = self.data[:]
        copied_queue.size = self.size
        return copied_queue

    def sort(self):
        """Sort the queue in ascending order."""
        self.data.sort()

    def slice(self, start, end):
        """Slice the queue and return a new queue."""
        if start < 0 or end > self.size or start >= end:
            raise RuntimeError("Invalid slice indices!")
        sliced_queue = Queue()
        sliced_queue.data = self.data[start:end]
        sliced_queue.size = end - start
        return sliced_queue

    def display(self):
        """Display the elements of the queue."""
        print(" ".join(map(str, self.data)))


# Main function for testing
if __name__ == "__main__":
    q1 = Queue()
    q2 = Queue()

    # Enqueue elements into q1
    q1.enqueue(10)
    q1.enqueue(20)
    q1.enqueue(30)

    # Enqueue elements into q2
    q2.enqueue(40)
    q2.enqueue(50)
    q2.enqueue(60)

    # Merge queues
    merged_queue = q1.merge(q2)
    print("Merged queue:", end=" ")
    merged_queue.display()

    # Reverse q1
    q1.reverse()
    print("Reversed q1:", end=" ")
    q1.display()

    # Copy q2
    copied_queue = q2.copy()
    print("Copied q2:", end=" ")
    copied_queue.display()

    # Sort q2
    q2.enqueue(25)
    q2.sort()
    print("Sorted q2:", end=" ")
    q2.display()

    # Slice q2
    sliced_queue = q2.slice(1, 4)
    print("Sliced q2 (1 to 4):", end=" ")
    sliced_queue.display()